man/man1/sdmath.1
.TH SDMATH 1NEMO "31 July 2025"

.SH "NAME"
sdmath \- SD math and file I/O benchmark

.SH "SYNOPSIS"
\fBsdmath\fP [parameter=value]

.SH "DESCRIPTION"
This program benchmarks some OpenMP parallelizable Single Dish vector math in a typical
4-phase calibration cycle with an ON/OFF with a separate COLD/HOT in each pointing.

.PP
Each spectrum (called row below) has \fBnchan\fP channels, Each \fBnscan\fP has 4 rows, with
the CalOn and CalOff (Hot and Cold if you wish)
for the Sig and the Ref (On and Off if you wish).
The following math is needed to construct a
spectrum, thus the total operations scales as nscan*nchan*iter.
.EX

    Tsys =  Tc * <row1> / <row2-row1>                 <cold> / <hot>-<cold>
    Ta = Tsys * ( (row1+row2)/(row3+row4) - 1)            on/off - 1
    Spectrum = <Ta>                                    time-averaging
    
.EE

This block of memory of size 4*nchan*nscan can be iteratively visited, to get a more
reliable measure of the CPU usage. It can make a huge difference if this block fits
in one of the CPU caches.

.PP

The benchmark here is very basic, and only deals with the math with simulated data without
the need to pass through a fitsio type library. The file I/O bench portion of this benchmark
reads a large file to get an idea of I/O overhead, even though many other tools exist
to measure disk I/O, from \fIhdparm(1)\fP to \fIiozone(1)\fP.
.PP
A slightly more realistic one can be found in \fIsdinfo(1NEMO)\fP
reading actual \fIsdfits(5NEMO)\fP files and selecting a set of operations.

.PP
The standard benchmark loops 10 times over 1000 scans of 100,000 channels (thus 1e9 row operations)
in about 2-3 secs. We call this 1 GRop.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBnscan=\fP\fIint\fP
Number of scans. Can be used to time-average (see \fBaver=\fP below).
.br
[1000]    
.TP
\fBnchan=\fP\fIint\fP
Number of channels per spectrum.
.br
[100000]    
.TP
\fBiter=\fP\fIint\fP
How many times to iterate (0 means do nothing).
.br
[10]
.TP
\fBmode=\fP\fIint\fP
Mode of math. Currently the equivalent of a 4-phase PS observation
is simulated, but a simpler 2-phase TP observation should be implemented as well.
mode=1:  random numbers   mode=2: sequence numbers, easier to compare with a python
implementation.
.br
[1]  
.TP
\fBaver=t|f\fP
Add time average over nscan.
.br
[f]  
.TP
\fBin=\fP\fIfilename\fP
If given, read this file into memory. Meant to measure disk I/O
.br
[Not used]
.TP
\fBbs=\fP\fIint\fP
Blocksize in kB to read file with.
.br
[16]
.TP
\fBmaxbuf=\fP\fIint\fP
Maximum buffer size (in MB) for reading buffer. If 0 is given, the full filesize is used,
which for large files can  potentially run into the swap.
.br
[0]
.TP
\fBseed=\fP\fIint\fP
seed for the random number generator. A value of 0 (the default)
will be converted into a unique new value using UNIX's clock time, in
seconds since 1970.0). See also \fIxrandom(1NEMO)\fP for more options.
.br
[Default: 0]


.SH "EXAMPLES"

Standard benchmark was extended to iter=100 to avoid system time. These are 10 Grops.
CPU times are user space
seconds [nemobench5 scores listed where available]
.EX

Xeon E5620  @ 2.40GHz (fourier) - 115.1  [218]
AMD EPYC 7302 @ 3.0 GHz (lma)   -  34.1  [675]
AMD Ryzen AI 9 HX PRO 370       -  23.4  [~1170]
Ultra 7 155H @ 4.5 GHz (d76)    -  21.8  [~1200]
M4 air                          -  15.4  [~2000]

.PP

.EE

To view the impact of OpenMP, here is an example performance on an Ultra 7 155H,
measured in wall clock time seconds, and iter=20:

.EX
OMP_NUM_THREADS=2 /usr/bin/time sdmath  1000 160000 iter=20


nscan  nchan      1      2      4      8     12     16     20 
----- ------    ---    ---    ---    ---    ---    ---    ---
 1000 160000    7.8    5.0    3.9    4.2    3.7    3.6    5.7
 2000  80000    7.8
 4000  40000    7.8 
 8000  20000    7.8
16000  10000    8.0    5.4    4.4    5.3    5.2    9.5    9.9
----- ------    ---    ---    ---    ---    ---    ---    ---

.EE
Somewhat surprisingly that barely a factor of 2 can be gained by going multi-core
on an typical laptop intel CPU.  An AMD CPU fares much better. The 
Ryzen AI 9 HX PRO 370 is able to lower the wall clock time all the way up to
12 cores.

.PP
To see the effect of the CPU cache, pick a case where nscan*nchan*iter is constant,
but
.EX
nscan nchan iter
1e1   1e1   1e7   1.52sec
1e1   1e2   1e6   1.42
1e1   1e3   1e5   1.77
1e1   1e4   1e4   1.82      3 MB
1e1   1e5   1e3   2.33     30 MB
1e1   1e6   1e2   2.52    300 MB
1e1   1e7   1e1   3.23   3000 MB
1e2   1e6   1e1   2.69   3000 MB
1e2   1e5   1e2   2.31    300 MB

.EE

.PP
A read benchmark can be done with a large series of dummy Plummer models. The example file created
here is about 5GB in size:
.EX
   mkplummer p1M 1000000 nmodel=100
   /usr/bin/time sdmath  in=p1M
   /usr/bin/time sdmath  in=p1M maxbuf=1000
.EE
the latter resets the buffer 5 times, where the last read is a partial one.

.SH "SEE ALSO"
sdinfo(1NEMO), fits(5NEMO)

.SH "FILES"
$NEMO/src/tutor/bench/sdmath.c

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
30-apr-2025	Created		PJT
.fi


src/tutor/bench/sdmath.c
/*
 * benchmark some SD : math and I/O
 *
 * The default nscan=1000 nchan=100000 iter=10
 * d76:    4.6s -> 3 Gop/sec128
 * lma:    7.3s    2
 * jansky: 3.4s    4
 * 
 *     30-apr-2025    Created  // PJT
 */
 
#include <nemo.h>


string defv[] = {
  "nscan=1000\n        Number of scans",
  "nchan=100000\n      Number of channels",
  "iter=10\n           How many times to iterate (0 means do nothing)",
  "mode=1\n            Mode to init the data (0=random 1=arange)",
  "aver=f\n            Add time average over nscan",
  "in=\n               Read a file into memory",
  "bs=16\n             Blocksize in kB to read",
  "maxbuf=0\n          Max buffer for file data in MB",
  "seed=0\n            Seed, if need be",
  "VERSION=1.5\n       5-may-2025",
  NULL,
};

string usage = "SD math bench";


//   assumed order:   on_cold, on_hot, off_cold, off_hot
//
//   Tsys =  Tc * <row1> / <row2-row1>      <cold> / <hot>-<cold>
//   on  = row1 + row2
//   off = row3 + row4
//   Ta = Tsys * (on/off - 1)
//   Spectrum = <Ta>


// n OPS
real sum1(int n, real *data)
{
  real sum = 0.0;
  #pragma omp parallel for reduction(+:sum)
  for (int i=0; i<n; i++) sum += data[i];
  return sum; 
}

// 2*n OPS
real sum2(int n, real *data1, real *data2)
{
  real sum = 0.0;
  #pragma omp parallel for reduction(+:sum)
  for (int i=0; i<n; i++) sum += data2[i]-data1[i];
  return sum; 
}




// this one went slower!!!
// 2*n OPS
real tsys2(int n, real *cold, real *hot)
{
  real a, b;
  // @todo these two can be done in parallel
#pragma omp parallel
  {
#pragma omp sections 
    {
#pragma omp section
      a = sum1(n, cold);
#pragma omp section
      b = sum2(n, cold, hot);
    }
  }
  return a/b;
}

real tsys(int n, real *cold, real *hot)
{
  static real tc = 10.0;
  // @todo these two can be done in parallel
  return tc * sum1(n, cold) / sum2(n, cold, hot);
}

void file_read(string fname)
{
  size_t nread, n = nemo_file_size(fname);
  char *data;
  size_t i, nreset=0;
  off_t o;
  int bs = getiparam("bs");   // "bs" in kB
  size_t bufsize = bs*1024;   
  size_t maxbuf = 1024*1024*getiparam("maxbuf"); // "maxbuf" was in MB

  if (maxbuf == 0) {
    data = allocate(n);
    maxbuf = n;
  } else
    data = allocate(maxbuf);

  printf("file size: %ld bytes ~ %g MB   bufsize=%ld ~ %g MB\n",n, (real)n/1024/1024, bufsize, (real)bufsize/1024/1024);
  printf("maxbuf:    %ld bytes = %d MB\n",maxbuf, getiparam("maxbuf"));
  
  stream fp = stropen(fname,"r");
  for(i=0, o=0; ; i++) {
    //o=0;
    nread = fread(&data[o], bufsize, 1, fp);
    if (nread<=0) break;
    o += bufsize;
    if (o >= maxbuf) {
      nreset++;
      o = 0;
      dprintf(1,"buffer RESET at i=%ld\n",i+1);
    } else
      dprintf(2,"buffer at i=%ld at o=%ld maxbuf=%ld\n",i+1,o,maxbuf);
  }
  strclose(fp);
  printf("Last block# %ld of %ld,  #resets %ld\n", i, bufsize, nreset);
}



void nemo_main(void)
{
  size_t nchan = getiparam("nchan");
  size_t nscan = getiparam("nscan");
  size_t iter = getiparam("iter");
  size_t n4 = nchan * nscan * 4;
  int i, j, mode = getiparam("mode");
  real *data, *row0, *row1, *row2, *row3, *row4;
  real *aver, *spec, t_onn;
  size_t ndata = n4 * sizeof(real);
  bool Qaver = getbparam("aver");
  int nphase = 4;

  if (hasvalue("in")) {
    file_read(getparam("in"));
    return;
  }

  data = (real *) allocate(n4 * sizeof(real));
  spec = (real *) allocate(nchan * sizeof(real));
  aver = (real *) allocate(nchan * sizeof(real));
  

  printf("data size: %ld bytes ~ %g MB  nscan=%ld nchan=%ld iter=%ld\n",
	 ndata, (real)ndata/1024/1024, nscan, nchan, iter);
  printf("# rops = %ld  ~ %g Grops\n",nchan*nscan*iter,  (nchan*nscan*iter)/1e9);
    

  // only random one scan, xrandom() is expensive
  if (mode==1) {
    int seed = init_xrandom(getparam("seed"));
    dprintf(1,"seed used = %d\n",seed);
    for (i=0; i<nphase*nchan; i++) data[i] = xrandom(0.0,1.0);
  } else {
    // simple 0,1,2 for reproducing python
    for (i=0; i<nphase*nchan; i++) data[i] = i;
  }
  // duplicate other rows, so we fill memory out of the caching
  row0 = data;
  row1 = data + nphase*nchan;
  for (j=1; j<nscan; j++) {
    for (i=0; i<nphase*nchan; i++) row1[i] = row0[i];
    row0 = row1;
    row1 = row0 + nphase*nchan;
  }
  
  while (iter--) {
    row0 = data;
    for (i=0; i<nscan; i++) {
      row1 = row0;
      row2 = row1 + nchan;
      row3 = row2 + nchan;
      row4 = row3 + nchan;
      row0 = row4;  // for the next iter
      // Tsys =  Tc * <cold> / <hot-cold>
      t_onn = tsys(nchan, row1, row2);          // 3*nchan

      #pragma omp parallel for
      for (j=0; j<nchan; j++) {
	spec[j] = t_onn*((row1[j]+row2[j])/(row3[j]+row4[j])-1);  // 4*nchan      3.90user 0.89system 0:04.79elapsed 99%CPU
	//spec[j] = ((row1[j]+row2[j])/(row3[j]+row4[j])-1);      // check-1      3.85user 0.90system 0:04.76elapsed 99%CPU
	//spec[j] = (row1[j]+row2[j])/(row3[j]+row4[j]);          // check-2      3.81user 0.88system 0:04.70elapsed 99%CPU 
	//spec[j] = row1[j]/row3[j];                              // check-3      3.77user 0.91system 0:04.69elapsed 99%CPU 
	//spec[j] = row1[j];                                      // check-4      3.37user 0.88system 0:04.26elapsed 99%CPU 
       	if (Qaver)
	  aver[j] = aver[j] + spec[j];
      } // j-chan
    } // i-scan
  } // iter

  dprintf(1,"data[0,1..last] = %g %g %g\n", data[0], data[1], data[n4-1]);
  dprintf(1,"sum=%g\n",sum1(nchan,aver));
  
}
